src/stdlib/RandomMeasures.jl

```diff
@@ -134,21 +134,42 @@ function _logpdf(d::SizeBiasedSamplingProcess{DirichletProcess{T}}, x::T) where
     return logpdf(Beta(one(T), d.rpm.α), x/d.surplus)
 end
 
-function _logpdf_table(d::DirichletProcess{V}, m::T) where {T<:AbstractVector{Int},V<:Real}
-    if sum(m) == 0
-        return zeros(V,1)
-    elseif any(m_ -> m_ == 0, m)
-        z = log(sum(m) - 1 + d.α)
-        K = length(m)
-        zid = findfirst(m_ -> m_ == 0, m)
-        lpt(k) = k == zid ? log(d.α) - z : log(m[k]) - z
-        return map(k -> lpt(k), 1:K)
-    else
-        z = log(sum(m) - 1 + d.α)
-        K = length(m)
-        lp(k) = k > K ? log(d.α) - z : log(m[k]) - z
-        return map(k -> lp(k), 1:(K+1))
+function _logpdf_table(d::DirichletProcess{T}, m::AbstractVector{Int}) where {T<:Real}
+    # compute the sum of all cluster counts
+    sum_m = sum(m)
+
+    # shortcut if all cluster counts are zero
+    iszero(sum_m) && return zeros(T, 1)
+
+    # pre-calculations
+    dα = d.α
+    z = log(sum_m) - 1 + dα
+    table_zero = log(dα) - z
+
+    # construct the table
+    K = length(m)
+    table = Vector{T}(undef, K)
+    contains_zero = false
+    @inbounds for i in 1:K
+        mi = m[i]
+
+        if iszero(mi)
+            if contains_zero
```

Comment:
Maybe add a comment here, so that people know why `table[i]` will be set to `-Inf`.